src/frontend/src/session/cursor.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use futures::StreamExt;
+use itertools::Itertools;
+use pgwire::pg_field_descriptor::PgFieldDescriptor;
+use pgwire::pg_response::StatementType;
+use pgwire::types::Row;
+use risingwave_common::session_config::QueryMode;
+use risingwave_sqlparser::ast::ObjectName;
+
+use super::{transaction, PgResponseStream};
+use crate::error::{ErrorCode, Result, RwError};
+use crate::handler::query::{distribute_execute, local_execute, BatchPlanFragmenterResult};
+use crate::handler::util::{to_pg_field, DataChunkToRowSetAdapter};
+use crate::session::SessionImpl;
+
+pub struct Cursor {
+    row_stream: PgResponseStream,
+    pg_descs: Vec<PgFieldDescriptor>,
+    remaining_rows: VecDeque<Row>,
+}
+
+impl Cursor {
+    pub async fn new(
+        plan_fragmenter_result: BatchPlanFragmenterResult,
+        session: Arc<SessionImpl>,
+    ) -> Result<Self> {
+        let (row_stream, pg_descs) = Self::create_stream(plan_fragmenter_result, session).await?;
+        Ok(Self {
+            row_stream,
+            pg_descs,
+            remaining_rows: VecDeque::<Row>::new(),
+        })
+    }
+
+    async fn create_stream(
+        plan_fragmenter_result: BatchPlanFragmenterResult,
+        session: Arc<SessionImpl>,
+    ) -> Result<(PgResponseStream, Vec<PgFieldDescriptor>)> {
```

Comment:
The code here seems to be very similar to query/execute(), what is the reason for not applying query/execute()

Reply:
I can extract a function to unify the similar part.

Reply:
updated